// atwork/atwork.js
"use strict"
var ticket = null;
var image = null;
var notify = null;
var token = null;
var param = null;
let user = null;
var getParam = function(arg) {
	arg.apiServer = arg.apiServer || "http://api.workplus.io"
	arg.apiServer = arg.apiServer.replace(new RegExp("/+$"), "");
	param = {
		apiServer: arg.apiServer,
		client_id: arg.client_id,
		client_secret: arg.client_secret,
		tenement_id: arg.tenement_id || "atwork",
		advanceTime: arg.advanceTime || 3600000,
		maxRetries: arg.maxRetries || 3
	};
	global.atworkCfg = param;
	return param;
};


var initModule = function(param) {
	token = require("./token")
	ticket = require("./ticket");
	image = require("./image");
	notify = require("./notify");
	user = require("./user");
};

var init = function(arg) {
	if (param) {
		return;
	} else if (!arg.client_id || !arg.client_secret) {
		throw new Error("请提供 client_id 和 client_secret");
	} else {
		initModule(getParam(arg));
	}
};

module.exports = function(arg) {
	init(Object.create(arg === undefined ? null : arg));
	return {
		validateUserTicket: ticket.validateUserTicket,
		getImage: image.getImage,
		uploadImage: image.uploadImage,
		notifyWithArticle: notify.notifyWithArticle,
		getAccessToken: token.getAccessToken,
		resetToken: token.resetToken,
		synGetToken: token.synGetToken,
		fuzzyQueryByAccountName: user.fuzzyQueryByAccountName
	};
}
