// atwork/notify.js
"use strict"

var request = require("request");
var getErrMesg = require("./util").getErrMesg;
var api = require("./api-server");
var tokenModule = require("./token");
var param = global.atworkCfg;


var notify = function(httpBody) {
	var getTokenAndNotify = function(success, fail) {
		var errors = function(e) {
			fail(e);
		}
		var notifyUser = function(token) {
			var arg = {
				method: 'POST',
				uri: param.apiServer + "/app/msg?access_token=" + token,
				json: true,
				encoding: 'utf-8',
				headers: {
					'Content-type': 'application/json'
				},
				body: httpBody
			};
			var callbacks = {
				fail: fail,
				success: success
			};
			console.info("消息推送的url是", arg.uri)
			api.getSendRequestFunc(arg, callbacks, param.maxRetries)();
		}
		tokenModule.getAccessToken().then(notifyUser, errors).catch(errors);
	}
	return new Promise(getTokenAndNotify)
}


var notifyWithArticle = function(params) {
	var httpBody = {
		type: "article",
		client_ids: params.users,
		body: {
			articles: [{
				"summary": params.description,
				"show_cover": true,
				"create_time": new Date().getTime(),
				"content_source": params.url,
				"author": "",
				"sort": 0,
				"title": params.title || "",
				"content": params.description || "",
				"url": params.url,
				"cover_url": params.coverUrl,
				"cover_media_id": params.coverMediaID
			}]
		}
	};
	console.log("消息推送的body是", require("util").inspect(httpBody));
	return notify(httpBody);
}

module.exports = {
	notifyWithArticle: notifyWithArticle
}
