// atwork/ticket.js
"use strict"
var util = require("./util").getErrMesg;
var api = require("./api-server");
var tokenModule = require("./token");
var param = global.atworkCfg;


var validateUserTicket = function(ticket, accountName, timeout) { //第三个参数实在太挫，仅仅是为了方便单元测试
	var validateWithToken = function(success, fail) {
		var error = function(e) {
			fail(e);
		}
		var validate = function(accessToken) {
			var arg = {
				method: "GET",
				uri: param.apiServer + "/ticket/" + ticket + "?access_token=" + accessToken
			};
			var callbacks = {
				checkResult: function(body) {
					return body.result.client_id === accountName;
				},
				fail: fail,
				success: success
			};
			console.info("验证ticket的url是", arg.uri)
			api.getSendRequestFunc(arg, callbacks, param.maxRetries, timeout)();
		};
		tokenModule.getAccessToken().then(validate).catch(error);
	}
	if (!ticket || !accountName) {
		throw new Error("ticket和accountName不能为空");
	} else {
		return new Promise(validateWithToken);
	}
};

module.exports = {
	validateUserTicket: validateUserTicket
}
